frontend/app/src/util/AppNavigation.ts

```diff
@@ -140,6 +149,128 @@ export class V1Strategy {
       ) ?? this.appPages[0]
     )
   }
+
+  clearPageElements(
+    _elements: AppRoot,
+    mainScriptHash: string,
+    sidebarElements: BlockNode | undefined
+  ): AppRoot {
+    return AppRoot.empty(mainScriptHash, false, sidebarElements)
+  }
+}
+
+export class V2Strategy {
+  readonly parent: AppNavigation
+
+  mainScriptHash: string | null
+
+  appPages: IAppPage[]
+
+  mainPage: IAppPage | null
+
+  hideSidebarNav: boolean | null
+
+  constructor(parent: AppNavigation) {
+    this.parent = parent
+    this.mainScriptHash = null
+    this.appPages = []
+    this.mainPage = null
+    this.hideSidebarNav = null
+  }
+
+  handleNewSession(newSession: NewSession): MaybeStateUpdate {
+    this.mainScriptHash = newSession.mainScriptHash
+    this.hideSidebarNav = newSession.config?.hideSidebarNav ?? false
+
+    // We do not know the page name, so use an empty string version
+    document.title = " · Streamlit"
+
+    return [{ hideSidebarNav: this.hideSidebarNav }, () => {}]
+  }
+
+  handlePagesChanged(_pagesChangedMsg: PagesChanged): MaybeStateUpdate {
+    // This message does not apply to V2
+    return undefined
+  }
+
+  handlePageNotFound(pageNotFound: PageNotFound): MaybeStateUpdate {
+    const { pageName } = pageNotFound
+    this.parent.onPageNotFound(pageName)
+
+    return [
+      { currentPageScriptHash: this.mainScriptHash ?? "" },
+      () => {
+        this.parent.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_CURRENT_PAGE_NAME",
+          currentPageName: "",
+          currentPageScriptHash: this.mainScriptHash ?? "",
+        })
+      },
+    ]
+  }
+
+  handleNavigation(navigationMsg: Navigation): MaybeStateUpdate {
+    const { sections, position, appPages } = navigationMsg
+
+    this.appPages = appPages
+    this.hideSidebarNav = this.hideSidebarNav || position === "hidden"
+
+    const currentPage = appPages.find(
+      p => p.pageScriptHash === navigationMsg.pageScriptHash
+    ) as IAppPage
+    const mainPage = appPages.find(p => p.isDefault) as IAppPage
+    this.mainPage = mainPage
+    const currentPageScriptHash = currentPage.pageScriptHash as string
+    const currentPageName = currentPage.isDefault
+      ? ""
+      : (currentPage.pageName as string).replaceAll(" ", "_")
+
+    document.title = `${currentPage.pageName as string} · Streamlit`
```

Comment:
nit: it might be a good idea to have a util function in this file to construct the title as this is the same as in [this line](https://github.com/streamlit/streamlit/pull/8673/files#diff-7f6abc01745f6d3e4805e1279acec8c843ba151febe680c576d1497a3fb5a077R84), e.g.

```typescript
function getTitle(pageName: string): string {
  return `${pageName as string} · Streamlit`
} 
```

and then call this^ in all places where we construct the document.title